projects/lights_control/src/lights_signal_fsm.c

```diff
@@ -0,0 +1,122 @@
+#include "lights_signal_fsm.h"
+
+// No signals are active.
+FSM_DECLARE_STATE(state_none);
+// Left signal active.
+FSM_DECLARE_STATE(state_left_signal);
+// Right signal active.
+FSM_DECLARE_STATE(state_right_signal);
+// Hazard button pressed.
+FSM_DECLARE_STATE(state_hazard_signal);
+// Hazard button pressed, while left signal active.
+FSM_DECLARE_STATE(state_hazard_left_signal);
+// Hazard button pressed, while right signal active.
+FSM_DECLARE_STATE(state_hazard_right_signal);
+
+static bool prv_guard_signal_left_on(const FSM *fsm, const Event *e, void *context) {
+  return (e->id == LIGHTS_EVENT_SIGNAL_ON) && (e->data == LIGHTS_EVENT_SIGNAL_MODE_LEFT);
+}
+
+static bool prv_guard_signal_left_off(const FSM *fsm, const Event *e, void *context) {
+  return (e->id == LIGHTS_EVENT_SIGNAL_OFF) && (e->data == LIGHTS_EVENT_SIGNAL_MODE_LEFT);
+}
+
+static bool prv_guard_signal_right_on(const FSM *fsm, const Event *e, void *context) {
+  return (e->id == LIGHTS_EVENT_SIGNAL_ON) && (e->data == LIGHTS_EVENT_SIGNAL_MODE_RIGHT);
+}
+
+static bool prv_guard_signal_right_off(const FSM *fsm, const Event *e, void *context) {
+  return (e->id == LIGHTS_EVENT_SIGNAL_OFF) && (e->data == LIGHTS_EVENT_SIGNAL_MODE_RIGHT);
+}
+
+static bool prv_guard_signal_hazard_on(const FSM *fsm, const Event *e, void *context) {
+  return (e->id == LIGHTS_EVENT_SIGNAL_ON) && (e->data == LIGHTS_EVENT_SIGNAL_MODE_HAZARD);
+}
+
+static bool prv_guard_signal_hazard_off(const FSM *fsm, const Event *e, void *context) {
+  return (e->id == LIGHTS_EVENT_SIGNAL_OFF) && (e->data == LIGHTS_EVENT_SIGNAL_MODE_HAZARD);
```

Comment:
What's the point of having separate on/off guards? The event ID is already guarded by the transition function itself.

Reply:
oh wow why'd I do that. :|